katran/lib/KatranLb.cpp

```diff
@@ -543,13 +542,68 @@ void KatranLb::loadBpfProgs() {
   progsLoaded_ = true;
   if (!config_.disableForwarding && features_.introspection) {
     startIntrospectionRoutines();
+    introspectionStarted_ = true;
   }
 
   if (!config_.disableForwarding) {
     attachLrus();
   }
 }
 
+bool KatranLb::reloadBalancerProg(
+    const std::string& path,
+    folly::Optional<KatranConfig> config) {
+  int res;
+  if (config_.disableForwarding) {
+    return false;
+  }
+
+  res = bpfAdapter_.reloadBpfProg(path);
+  if (res) {
+    return false;
+  }
+
+  if (config.has_value()) {
+    config_ = *config;
+  }
+
+  config_.balancerProgPath = path;
+
+  initialSanityChecking();
+  featureDiscovering();
+
+  if (features_.gueEncap) {
+    setupGueEnvironment();
+  }
+
+  if (features_.inlineDecap) {
+    enableRecirculation();
+  }
+
+  // add values to main prog ctl_array
+  std::vector<uint32_t> balancer_ctl_keys = {kMacAddrPos};
+
+  for (auto ctl_key : balancer_ctl_keys) {
+    res = bpfAdapter_.bpfUpdateMap(
+        bpfAdapter_.getMapFdByName("ctl_array"),
+        &ctl_key,
+        &ctlValues_[ctl_key]);
+
+    if (res != 0) {
+      throw std::invalid_argument(folly::sformat(
+          "can't update ctl array for main program, error: {}",
+          folly::errnoStr(errno)));
+    }
+  }
+
+  if (features_.introspection && !introspectionStarted_) {
+    startIntrospectionRoutines();
+    introspectionStarted_ = true;
```

Comment:
Udip noticed: should we reset `introspectionStarted_` if `!features_.introspection`?

Reply:
if introspection is not enabled why should we reset anything? this flag is used for us to start KatranMonitor only once. so e.g. if 1st version  of prog contains introspection and 2nd does not - we will not stop KatranMonitro. it wont just receive any data.

Reply:
Ok, that makes sense. Nit: perhaps we should consider make things explicit by use the words "introspection" and "monitor" more consistently, because the former is more in the bpf side of things and the later is more of client-facing concept.